geopmctlpy/geopmctl/metric.py

```diff
@@ -0,0 +1,680 @@
+#
+#      * Redistributions in binary form must reproduce the above copyright
+#        notice, this list of conditions and the following disclaimer in
+#        the documentation and/or other materials provided with the
+#        distribution.
+#
+#      * Neither the name of Intel Corporation nor the names of its
+#        contributors may be used to endorse or promote products derived
+#        from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+#  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+'''
+Defines Metric/Signal related classes and factory method(s).
+'''
+
+from enum import Enum
+from abc import ABC
+from abc import abstractmethod
+import yaml
+from geopmdpy import pio
+from geopmdpy import topo
+
+
+def create_signal(descriptor, report_formatter=None, trace_formatter=None):
+    '''
+    Factory function for creating a Signal object.
+
+    Args
+
+    descriptor (tuple of str, int, int): Tuple of name of the GEOPM signal, a domain enumeration
+    from the geopmdpy.topo package and instance number.
+    Example: ("POWER_PACKAGE", geopmdpy.topo.DOMAIN_BOARD, 0)
+
+    report_formatter (Formatter): Formatter for outputting the value of accumulator in a report
+    file. Passed to the Metric instance created from this signal as sample_only.
+
+    trace_formatter (Formatter): Formatter for outputting the metric sample value in a trace
+    file. Passed to the Metric instance created from this signal as sample_only.
+    '''
+    return_val = None
+    behavior = Behavior(pio.signal_info(descriptor[0])[2])
+    if behavior is Behavior.MONOTONE:
+        return_val = MonotoneSignal(descriptor, report_formatter=report_formatter,
+                                    trace_formatter=trace_formatter)
+    else:
+        return_val = Signal(descriptor, behavior, report_formatter=report_formatter,
+                            trace_formatter=trace_formatter)
+    return return_val
+
+
+class Behavior(Enum):
+    '''
+    Enum representing Signal behaviors. Should match the integer representation of behaviors
+    returned by pio.
+    '''
+    CONSTANT = 0
+    MONOTONE = 1
+    VARIABLE = 2
+    LABEL = 3
+
+
+class SignalList:
```

Comment:
This seems to be a helper object for the Agents.  This should go in the runtime.py if that's the case.